### PR TITLE
Bnb/exo refactor

### PR DESCRIPTION
**Pull request by bnb32**
Changed exogenous data extraction to use a dictionary specifying model and combine_type (input/layer/output) for each use of exogenous data. Refactored all uses of exo data in models/tests/fwp according to this format. 

Added SZA exogenous data handler for future use in solar models. 

Combined basic gan models and wind gan models, including conditional models, which revert to basic models without exogenous data input.

### CODE REVIEW

File: sup3r/models/multi_step.py [hunk 1]
```diff
@@ -131,6 +112,82 @@ def seed(s=0):
...
+            model_step
+        """
+        model_step_exo = None
+        if exogenous_data is not None:
+            model_step_exo = {}
+            for feature in exogenous_data:
+                steps = [step for step in exogenous_data[feature]['steps']
+                         if step['model'] == model_step]
+                if steps:
+                    model_step_exo[feature] = {'steps': steps}
```
Comment (grantbuster) [inline sup3r/models/multi_step.py:149]: It might be worth defining a separate data structure for the `exogenous_data` dictionary, especially if it defines the data structure more explicitly thereby helping users decipher how to set up the config. 

Consider a couple of classes: `class SingleExoDataStep(combine_type, model, data)`, `class ExoData(steps)`... Would be extra good if `ExoData` could be init from a config file directly OR a collection of `SingleExoDataStep` instances. Then the `ExoData` class would have methods to retrieve various bits of the data. 

Not totally sure how this would fit into the refactored fwp config as i haven't gotten there yet.

File: sup3r/models/multi_step.py [hunk 2]
```diff
@@ -320,6 +374,57 @@ def load(cls, spatial_model_dirs, temporal_model_dirs, verbose=True):
 
         return cls(s_models, t_models)
 
+    def _split_exo_dict(self, split_step, exogenous_data=None):
```
Comment (grantbuster) [inline sup3r/models/multi_step.py:377]: I haven't looked far enough to know the answer to this, but shouldn't you not have to split the exo_data dict into spatial/temporal parts now that each entry has an explicitly defined model step? Like wouldn't the extra model steps just be passed through and unused?

File: sup3r/pipeline/forward_pass.py [hunk 3]
```diff
@@ -693,7 +693,14 @@ def __init__(
...
             exogenous features such as topography for future multistep foward
-            pass
+            pass. This should be a nested dictionary with keys for each
+            exogeneous feature. The dictionaries corresponding to the feature
+            names should include the path to exogenous data source, the
+            resolution of the exogenous data, and how the exogenous data should
+            be used in the model. e.g. {'topography': {'file_paths': 'path to
+            input files', 'source_file': 'path to exo data', 'exo_resolution':
+            {'spatial': '1km', 'temporal': None}, 'steps': [{'model': 0,
+            'combine_type': 'input'}, {'model': 0, 'combine_type': 'layer'}]}
```
Comment (grantbuster) [inline sup3r/pipeline/forward_pass.py:703]: Related to a previous comment, there's got to be a better way to specify this input. Could we more rigorously define the `ExoExtract` or `ExogenousDataHandler` class to be initialized directly on this input? These kwargs basically get sent to `ExogenousDataHandler` don't they? Let's just put all the parsing into the class and then in this docstring we can just say go look at the args required by the class.

File: sup3r/pipeline/forward_pass.py [hunk 4]
```diff
@@ -1128,13 +1113,210 @@ def __init__(self, strategy, chunk_index=0, node_index=0):
...
+            Full exo_kwargs dictionary with all feature entries.
+            e.g. {'topography': {'exo_resolution': {'spatial': '1km',
+            'temporal': None}, 'steps': [{'model': 0, 'combine_type': 'input'},
+            {'model': 0, 'combine_type': 'layer'}]}}
+
+        Returns
+        -------
+        updated_dict : dict
+            Same as input dictionary with s_agg_factor, t_agg_factor,
+            s_enhance, t_enhance added to each step entry for all features
```
Comment (grantbuster) [inline sup3r/pipeline/forward_pass.py:1266]: FWIW i would still prefer to input the agg factors explicitly so we can tweak them as a sensitivity and have less code logic.

File: sup3r/preprocessing/data_handling/exo_extraction.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+        return source_lat_lon
+
+
+class SzaExtract(ExoExtract):
+    """SzaExtract for H5 files"""
+
+    @property
+    def source_data(self):
+        """Get the 1D array of sza data from the exo_source_h5"""
+        return SolarPosition(self.hr_time_index,
```
Comment (grantbuster) [inline sup3r/preprocessing/data_handling/exo_extraction.py:469]: nice thanks for implementing this! Wait, no way this is a 1D array though, right?

**Review by grantbuster — CHANGES_REQUESTED** [review]
What a burly PR! Overall i really like what you've done to bring the exo data handing into the base GAN model architecture. I think the config structure of the exo_data stuff makes sense, my main comment is just to put all of the initialization logic into a class and then point to that class for how to setup a valid config. Putting these special endless kwarg data structures into separate classes will make sure we have a dedicated place for documentation and an easier time with future developments.

File: sup3r/models/multi_step.py (hunk 1 above)
Comment (bnb32) [reply sup3r/models/multi_step.py:149]: This is a cool idea. Maybe better to leave for another PR so this initial big refactor can be merged first?

**Comment by bnb32** [comment]
> What a burly PR! Overall i really like what you've done to bring the exo data handing into the base GAN model architecture. I think the config structure of the exo_data stuff makes sense, my main comment is just to put all of the initialization logic into a class and then point to that class for how to setup a valid config. Putting these special endless kwarg data structures into separate classes will make sure we have a dedicated place for documentation and an easier time with future developments.

Moved most of the dict parsing methods from forward_pass into ExogenousDataHandler. Do you think this (and methods in model classes) could be improved with data structure classes?

File: sup3r/models/abstract.py [hunk 6]
```diff
@@ -66,27 +67,6 @@ def seed(s=0):
         """
         CustomNetwork.seed(s=s)
 
-    @abstractmethod
-    def generate(self, low_res):
```
[the remark this answers is not in the document]
Comment (grantbuster) [reply sup3r/models/abstract.py:70]: From AbstractSingleModel? The linear models only inherit from AbstractInterface and so it could be worthwhile to keep the abstract generate method here.

File: sup3r/models/abstract.py [hunk 7]
```diff
@@ -106,17 +86,275 @@ def input_dims(self):
...
+        return self.meta.get('input_resolution', None)
+
+    def _get_numerical_resolutions(self):
+        """Get the input and output resolutions without units"""
+        ires_num = {k: int(re.search(r'\d+', v).group(0))
+                    for k, v in self.input_resolution.items()}
+        enhancements = {'spatial': self.s_enhance,
+                        'temporal': self.t_enhance}
+        ores_num = {k: v // enhancements[k] for k, v in ires_num.items()}
+        return ires_num, ores_num
```
[the remark this answers is not in the document]
Comment (grantbuster) [reply sup3r/models/abstract.py:150]: I'd still clarify this in the docstring of `_get_numerical_resolutions`. You say "without units" but that doesnt specify the type. Just add `e.g., {'spatial': 30, 'temporal': 60} for 30km 60min`

File: sup3r/models/abstract.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bnb32) [reply sup3r/models/abstract.py:70]: Oh yeah fair.

File: sup3r/models/abstract.py [hunk 8]
```diff
@@ -106,17 +86,275 @@ def input_dims(self):
...
+    @property
+    def output_resolution(self):
+        """Resolution of output data. Given as a dictionary {'spatial':...,
+        'temporal':...}"""
+        output_res = self.meta.get('output_resolution', None)
+        if self.input_resolution is not None and output_res is None:
+            output_res = self.input_resolution.copy()
+            ires_num, ores_num = self._get_numerical_resolutions()
+            output_res = {k: v.replace(str(ires_num[k]), str(ores_num[k]))
+                          for k, v in self.input_resolution.items()}
```
[the remark this answers is not in the document]
Comment (grantbuster) [reply sup3r/models/abstract.py:201]: Gotcha. I think we should probably protect against something weird like input/output resolution being `44km` and `11km` and this getting mapped to `1111km` due to the double replacement. Probably a rare input but could definitely happen.

File: sup3r/models/abstract.py [hunk 9]
```diff
@@ -106,17 +86,275 @@ def input_dims(self):
...
+        if check:
+            exo_data = {k: v for k, v in exogenous_data.items()
+                        if k in self.training_features}
+            for i, (feature, entry) in enumerate(exo_data.items()):
+                f_idx = low_res_shape[-1] + i
+                training_feature = self.training_features[f_idx]
+                msg = ('The ordering of features in exogenous_data conflicts '
+                       'with the ordering of training features. Received '
+                       f'{feature} instead of {training_feature}.')
+                assert feature == training_feature, msg
```
[the remark this answers is not in the document]
Comment (grantbuster) [reply sup3r/models/abstract.py]: You implemented here right? https://github.com/NREL/sup3r/pull/167/commits/0e874c63d4c2149d8e6212d802e4e632bc915ae7#diff-96aba2bb75412f04eccfe154683dac43c4cb3459244158a0d591b4e59a1ccb5dR244

File: sup3r/models/abstract.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bnb32) [reply sup3r/models/abstract.py:201]: Why would that happen? ires_num would be 44 ores_num would be 11 and the replacement would replace 44 will 11 no?

File: sup3r/models/abstract.py [hunk 10]
```diff
@@ -106,17 +86,275 @@ def input_dims(self):
...
+        -------
+        high_res_gen : tf.Tensor
+            Same as input with exogenous data combined with high_res input
+        """
+        if high_res_true.shape[-1] > high_res_gen.shape[-1]:
+            for feature in self.exogenous_features:
+                f_idx = self.training_features.index(feature)
+                exo_data = high_res_true[..., f_idx: f_idx + 1]
+                high_res_gen = tf.concat((high_res_gen, exo_data), axis=-1)
+        return high_res_gen
```
[the remark this answers is not in the document]
Comment (grantbuster) [reply sup3r/models/abstract.py:327]: Very nice way to generalize this functionality

File: sup3r/models/abstract.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bnb32) [reply sup3r/models/abstract.py]: Yeah exactly. Forgot to reply to this one.

File: sup3r/models/abstract.py [hunk 11]
```diff
@@ -106,17 +86,275 @@ def input_dims(self):
...
+
+    def _get_exo_val_loss_input(self, high_res):
+        """Get exogenous feature data from high_res
```
[the remark this answers is not in the document]
Comment (grantbuster) [reply sup3r/models/abstract.py]: We still have "loss" in the name. Is this intentional? This method seems generic enough to not be specific to any loss calculation? Might be missing something

File: sup3r/models/abstract.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (bnb32) [reply sup3r/models/abstract.py:327]: Thanks!

File: sup3r/models/multi_step.py (hunk 2 above)
Comment (grantbuster) [reply sup3r/models/multi_step.py:377]: Hmmm yeah we talked about this right? Getting rid of the spatial+temporal model class? Sounding like a better and better idea! Flat is better than nested, right?

File: sup3r/models/abstract.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (bnb32) [reply sup3r/models/abstract.py]: It's only in the case of computing loss that we take the exo data from the high res array. For instance, in the case of input to the fwp we get the exo data from the exo_data dictionary

File: sup3r/models/abstract.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (grantbuster) [reply sup3r/models/abstract.py:201]: mmmm yeah i think you're right. not sure what i was thinking. nevermind!

File: sup3r/models/multi_step.py (hunk 2 above)
Comment (bnb32) [reply sup3r/models/multi_step.py:377]: Yep! Removed the spatial+temporal gan models but kept the Base class and changed the Solar model to subclass that base class.

File: sup3r/models/abstract.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (grantbuster) [reply sup3r/models/abstract.py]: Yeah a good name should be descriptive of what the function does though, not necessarily it's intended use case. I could see me wanting to use it in a test script to retrieve data from an exo data dictionary. This isn't super important but I think the name would be better without the "loss" specificity.

File: sup3r/models/multi_step.py (hunk 1 above)
Comment (grantbuster) [reply sup3r/models/multi_step.py:149]: Yeah that's fine, i think separate data struct classes and removing the spatial/temporal model are the two pieces left "todo", right?

Comment (bnb32) [reply sup3r/models/multi_step.py:149]: Just the struct classes, if were ok with keeping the SpatialThenTemporalBase used only by the Solar model.

File: sup3r/models/abstract.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (bnb32) [reply sup3r/models/abstract.py]: Yeah thats a good point.

**Review by grantbuster — APPROVED** [review]
